[#21963] YSQL: Configure clockbound agent (#24434)
yugabyte/yugabyte-db: docs/content/preview/deploy/manual-deployment/system-config.md

## Set up time synchronization

YugabyteDB relies on clock synchronization to guarantee consistency in distributed transactions. Chrony is the prefered NTP implementation for clock synchronization.

### Install Chrony

To install chrony, run:

```sh
$ sudo yum install -y chrony
```

### Configure Precision Time Protocol

{{<tags/feature/tp>}} Precision Time Protocol (PTP) is a network protocol designed for highly accurate time synchronization across devices in a network. PTP provides microsecond-level accuracy. PTP relies on a PTP Hardware Clock (PHC), a dedicated physical clock device that enhances time synchronization accuracy.

Currently, PTP is only available for AWS. To check if your AWS instance supports PTP and PHC, see [AWS PTP Hardware Clock](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configure-ec2-ntp.html#connect-to-the-ptp-hardware-clock).

Configure PTP using the `configure_ptp.sh` script in the bin directory of your YugabyteDB home directory as follows:

```sh
sudo bash ./bin/configure_ptp.sh
```

### Configure ClockBound

> {{<tags/feature/tp>}} [ClockBound](https://github.com/aws/clock-bound) is an open source daemon that allows you to compare timestamps to determine order for events and transactions, independent from the instances’ respective geographic locations. ClockBound provides a strict interval in which the reference time (true time) does exist.

Reviewer: ```suggestion
{{<tags/feature/tp>}} [ClockBound](https://github.com/aws/clock-bound) is an open source daemon that allows you to compare timestamps to determine order for events and transactions, independent of an instance's geographic location. ClockBound provides a strict interval in which the reference time (true time) exists.
```

Author: Done

Added a small change to read-restart-error.md as well.

Although optional, configuring ClockBound improves clock accuracy by several orders of magnitude. ClockBound requires chrony and can be used in conjunction with PTP.

Configure ClockBound using the `configure_clockbound.sh` script in the bin directory of your YugabyteDB home directory as follows:

```sh
sudo bash ./bin/configure_clockbound.sh
```

After configuring ClockBound, you must configure the [YB-TServer](../start-tservers/) and [YB-Master](../start-masters/) servers with the `time_source=clockbound` flag.

If the ClockBound agent is configured with PTP, use a more aggressive clock error estimate such as `clockbound_clock_error_estimate_usec=100`.

> #### Verify ClockBound configuration

Reviewer 2: ```suggestion
### Verify ClockBound configuration
```

Reviewer 2: This can be just a H3

Author: Done

## ntp

> YugabyteDB is a distributed database and relies on clock synchronization to

Reviewer 2: Can we  say :  YUgabyteDB relies on Clock synchronization to guarantee consistency in distributed transactions

Author: Done

$ sudo yum install -y chrony
```

> ### PHC (PTP Hardware Clock) {{<tags/feature/tp>}}

Author: Done

> PTP Hardware Clock (PHC) refers to a physical clock device that supports the [Precision Time Protocol (PTP)](https://en.wikipedia.org/wiki/Precision_Time_Protocol) which is a network protocol used for highly accurate time synchronization.

Author: Done

> To check if your AWS instance supports PHC, see [AWS PTP Hardware Clock](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configure-ec2-ntp.html#connect-to-the-ptp-hardware-clock).

Author: PHC is only supported on AWS. No other cloud providers are supported at the moment.

> To check if your AWS instance supports PHC, see [AWS PTP Hardware Clock](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configure-ec2-ntp.html#connect-to-the-ptp-hardware-clock).

Author: Added a small comment clarifying that.

> Configure PTP using

Author: Done

```sh
sudo bash ./bin/configure_ptp.sh
```

> ### Clockbound Agent {{<tags/feature/tp>}}

Author: Done

> Configuring a clockbound agent improves clock accuracy significantly.

Author: > so PHC is optional for any installation? for ntp only?
1. The above recommendation to install the ntp package is inaccurate. Users must install chrony whenever possible (and CentOS7 is End-Of-Life anyway).
2. You can configure clockbound with or without PHC.

> chrony is prerequisite for clockbound agent only?
However, you need chrony for configuring clockbound. So, this is a pre-requisite for clockbound.

Added a small phrase saying

pre-requisite for configuring clockbound.

> Configuring a clockbound agent improves clock accuracy significantly.

Author: Done

Eventually, everyone should configure it when this feature becomes GA. But we can incorporate that later on.

> Configuring a clockbound agent improves clock accuracy significantly.

Reviewer: What is clockbound agent?

> Configuring a clockbound agent improves clock accuracy significantly.

Reviewer: Should we just change the ntp section to chrony only?
Do PHC and clockbound belong as subsections of ntp/chrony, or should they be at the same level?

> Configuring a clockbound agent improves clock accuracy significantly.

Author: > What is clockbound agent? Is it https://github.com/aws/clock-bound?
Yes.

> Is it also AWS only?
No. It requires chrony daemon though.

> Configuring a clockbound agent improves clock accuracy significantly.

Author: > Should we just change the ntp section to chrony only?

Done.
Additional context: https://shorturl.at/dIdwt strongly recommends using chrony for example.

> Do PHC and clockbound belong as subsections of ntp/chrony, or should they be at the same level?

This is very subjective. What prompts you to ask this question? I prefer to keep this under ntp/chrony since PHC and clockbound are an extension of time synchronization. In fact, I would've named this section as time sync. Time sync is strictly necessary for distributed transactions which is at the heart of YugabyteDB.

While PHC is beneficial, it is not required. However, Chrony is a necessary prerequisite.

> 

Author: Done (minor typos resolved).

```sh
sudo bash ./bin/configure_clockbound.sh
```

> The scripts are packaged in YugabyteDB releases.

Author: Done

> Please configure yb-tserver and yb-master with gFlag `time_source=clockbound`.

Author: Done

I dont think the user wil reach this step without configuring clockbound. I can change it to

After configuring a clockbound agent ...

$ sudo yum install -y chrony
```

### PHC (PTP Hardware Clock)

**Supported Platforms:** AWS

{{<tags/feature/tp>}} PTP Hardware Clock (PHC) refers to a physical clock device that supports the [Precision Time Protocol (PTP)](https://en.wikipedia.org/wiki/Precision_Time_Protocol), which is a network protocol used for highly accurate time synchronization.

To check if your AWS instance supports PHC, see [AWS PTP Hardware Clock](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configure-ec2-ntp.html#connect-to-the-ptp-hardware-clock).

> Configure PTP using the `configure_ptp.sh` script in the bin directory of your YugabyteDB home directory as follows:

Reviewer: Are we configuring PTP or PHC? Is there a distinction here? They appear to be separate, but the section is called PHC not PTP

Author: > Are we configuring PTP or PHC?
Good question. We are doing both.

Basically the steps involve
1. Installing a new network driver in the kernel with PHC enabled.
2. Changing chrony to use PHC as a time source and PTP for time sync.

To clarify,
1. time source represents a physical clock (hardware clock) that is giving us timestamp readings (PHC)
2. time sync is making sure our clock is reading the same values as the rest of the world (PTP)

> Is there a distinction here?
Yes.

PHC is a hardware clock.
1. The device must support it (the link indicates which hardware supports this).
2. The kernel must support it (we compile the network driver with PHC support and install it).

PTP is a network protocol used for time synchronization.
PTP decides what messages goes over the network and how the hardware clock (PHC) is adjusted. PTP requires PHC to be available.

Both are configured using the configure_ptp.sh script.

> They appear to be separate, but the section is called PHC not PTP

I've changed the section a little bit to clarify things. Let me know if this is better.

$ sudo yum install -y chrony
```

### PHC (PTP Hardware Clock) {{<tags/feature/tp>}}

> Hardware clocks are more accurate.

Reviewer 2: PTP Hardware Clock (PHC) refers to a physical clock device that supports the [Precision Time Protocol (PTP)](https://en.wikipedia.org/wiki/Precision_Time_Protocol) which is used for highly accurate time synchronization in computer networks. These hardware clocks are more accurate.

Author: Done

> See https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configure-ec2-ntp.html#connect-to-the-ptp-hardware-clock

Reviewer 2: To check if your AWS instance supports PHC, see [AWS PTP Hardware clock](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configure-ec2-ntp.html#connect-to-the-ptp-hardware-clock)

Author: Done

to check whether your AWS instance supports PHC.

Configure PTP using

```sh
sudo bash ./bin/configure_ptp.sh
```

### Clockbound Agent {{<tags/feature/tp>}}

Configuring a clockbound agent improves clock accuracy significantly.
PHC is desirable but not mandatory.

> Chrony is a pre-requisite.

Reviewer 2: While PHC is beneficial, it is not required. However, Chrony is a necessary prerequisite.

Author: Done
